### DISCUSSION TITLE
Adding magma as a new language

### DISCUSSION THREAD

**Discussion by roed314** [github-linguist/linguist#6583] [Classification] [answered]
[Magma](http://magma.maths.usyd.edu.au/magma/) is a computer algebra system that is widely used by research mathematicians in number theory, group theory and other parts of algebra.  The conventional file extension is `.m`, which conflicts with many other languages, and leads to misclassification (unsurprisingly, since it's not currently in linguist).  I was curious about how difficult it would be to add, and had a few questions about doing so.
* [Your documentation](https://github.com/github-linguist/linguist/blob/master/CONTRIBUTING.md) suggests 200 repos as a threshold for a language being widespread.  With a file extension like `.m`, how can I check for that condition being satisfied?
* The language is closed source, and I don't know of an existing grammar.  Is a grammar a requirement for adding a language?  I assume that writing a grammar is not trivial, but can you point me at documentation for creating one?

Thanks!

**Comment by lildude** [comment, marked as answer]
> [Your documentation](https://github.com/github-linguist/linguist/blob/master/CONTRIBUTING.md) suggests 200 repos as a threshold for a language being widespread. With a file extension like .m, how can I check for that condition being satisfied?

We use the standard GitHub search and more [relaxed requirements](https://github.com/github-linguist/linguist/issues/5756) at the moment.

> The language is closed source, and I don't know of an existing grammar. Is a grammar a requirement for adding a language? I assume that writing a grammar is not trivial, but can you point me at documentation for creating one?

A grammar isn't a requirement but it kinda defeats the point adding support without one as most people only want support so they get syntax highlighting. We want a standard TextMate compatible grammar so any of the many TextMate grammar docs you can find with a simple search should get you on the right path.

As for the conflicts… this can be resolved with a heuristic which is just a regex that identifies something unique to the language, ideally within the first few lines of the file.

**Reply by roed314** [reply]
Thanks!
